lightly/models/utils.py

```diff
@@ -27,6 +27,75 @@
     from timm.models.vision_transformer import VisionTransformer
 
 
+def masked_pooling(
+    source: Tensor, mask: Tensor, reduce: str = "mean", num_cls: Optional[int] = None
+) -> Tensor:
+    """Reduce image feature maps (B x C x H x W) or (C x H x W) according to an integer
+    index given by `mask` (B x H x W) or (H x W).
+
+    Args:
+        source: Float tensor of shape (B x C x H x W) or (C x H x W) to be reduced.
+        mask: Integer tensor of shape (B x H x W) or (H x W) containing the integer indices.
+        reduce: The reduction operation to be applied, one of 'prod', 'mean', 'amax' or
+            'amin'. Defaults to 'mean'.
+        num_cls: The number classes in the possible masks. If None, the number of classes
+            is inferred from the unique elements in `mask`. This is useful when not all
+            classes are present in the mask.
+
+    Returns:
+        A tensor of shape (B x C x N) or (C x N) where N is the number of unique elements
+        in `mask` or `num_cls` if specified.s
+    """
+    if source.dim() == 3:
+        return _mask_reduce(source, mask, reduce, num_cls)
+    elif source.dim() == 4:
+        return _mask_reduce_batched(source, mask, num_cls)
+    else:
+        raise ValueError("source must have 3 or 4 dimensions")
+
+
+def _mask_reduce(
+    source: Tensor, mask: Tensor, reduce: str = "mean", num_cls: Optional[int] = None
+) -> Tensor:
```

Comment:
I was actually orienting myself at `torch.nn` layers e.g. `Conv2d` and `MaxPool2d` which also support batched and non-batched inputs. However, from what I have seen now, the functional API that they call also only works on batched inputs.

What do you think about kicking the unbatched implementation out, but creating an additional `nn.Module` wrapper `MaskedPooling` that works with unbatched and batched inputs? This would then truly follow the torch conventions.

Reply:
kept the unbatched version, but now calling the batched one from within there as you suggested: https://github.com/lightly-ai/lightly/pull/1739/commits/b5ba0b94bb3c00e8fa3b24138287ac512847c93b